reverse_proxy_handler.cc

```diff
@@ -1,22 +1,185 @@
+              } else {
+                  printf("\"remote_host\" statement in config needs more tokens\n");
+              }
+          }
+          if (config.statements[i]->tokens[0] == "remote_port") {
+              // remote_port must have 2 tokens
+              if (config.statements[i]->tokens.size() > 1) {
+                  remote_port = config.statements[i]->tokens[1];
+              } else {
+                  printf("\"remote_port\" statement in config needs more tokens\n");
+              }
+          }
+      }
+    }
+    if (!remote_host.empty() && !remote_port.empty()) {
+        original_uri_prefix = uri_prefix;
+        return RequestHandler::OK;
+    } else {
+        printf("Missing remote_port or remote_host provided to ReverseProxyHandler\n");
+        return RequestHandler::Error;
+    }
 }
 
 // Handles an HTTP request, and generates a response. Returns a response code
 // indicating success or failure condition. If ResponseCode is not OK, the
 // contents of the response object are undefined, and the server will return
 // HTTP code 500
 RequestHandler::Status ReverseProxyHandler::HandleRequest(const Request& request,
-Response* response) {
-    // TODO
-    return RequestHandler::OK;
-}
+    Response* response) {
+    boost::asio::io_service io_service;
+    tcp::socket socket(io_service);
+    tcp::resolver resolver(io_service);
+    tcp::resolver::query query(remote_host, remote_port);
+
+    // Use resolver to handle DNS lookup if query is not an IP address
+    boost::system::error_code ec;
+
+    // A hostname could resolve to multiple endpoints to try;
+    // connect() will try all of them
+    // See: http://www.boost.org/doc/libs/1_62_0/boost/asio/connect.hpp
+    boost::asio::connect(socket, resolver.resolve(query), ec);
+    if (ec == boost::asio::error::not_found) {
+        printf("Reverse proxy connection attempt to remote host failed. Check hostname and port number.\n");
+        return RequestHandler::Error;
+    }
+
+    std::cerr << "Got past connecting to remote_host!" << std::endl;
+
+    // Construct new request to send to remote host
+    // Example: Modify request from: /reverse_proxy/static/file1.txt
+    // to:                           /static/file1.txt
+    std::string remote_uri = request.uri();
+    remote_uri.erase(0, original_uri_prefix.size());
+
+    if (remote_uri.empty()) {
+      remote_uri = "/";
+    }
+
+    std::string remote_request
+      = "GET " + remote_uri + " HTTP/1.1" + "\r\n\r\n";
+
+    std::cerr << "Sending remote_request: " << remote_request << std::endl;
+
+    boost::asio::write(socket, boost::asio::buffer(remote_request));
 
+    // Based on: https://github.com/UCLA-CS130/Mr.-Robot-et-al./blob/c9b064c68cd4bc1ae6b5c012db59eae9cb8b946d/lightning_server.cc#L26
+    const int MAX_BUFFER_LENGTH = 8192;
+    char response_buffer[MAX_BUFFER_LENGTH];
+    memset(response_buffer, 0, MAX_BUFFER_LENGTH);
 
+    socket.read_some(boost::asio::buffer(response_buffer), ec);
```

Comment:
may want to put a loop here eventually for when responses exceed 8192 (for instance, when retrieving an image)